expr/func_aspercent.go

```diff
@@ -0,0 +1,241 @@
+			totalSeries = getTotalSeries(metaSeries)
+			// calculate sum of totals series
+		} else if totals != nil {
+			totalSeriesLists := groupSeriesByKey(totals, s.nodes, &keys)
+			totalSeries = getTotalSeries(totalSeriesLists)
+		} else {
+			return nil, errors.New("total must be None or a seriesList")
+		}
+
+		for key := range keys {
+			// No series for total series
+			if _, ok := metaSeries[key]; !ok {
+				serie2 := totalSeries[key]
+				serie2.QueryPatt = fmt.Sprintf("asPercent(MISSING,%s)", serie2.QueryPatt)
+				serie2.Target = fmt.Sprintf("asPercent(MISSING,%s)", serie2.Target)
+				serie2.Tags = map[string]string{"name": serie2.Target}
+				for i := range serie2.Datapoints {
+					serie2.Datapoints[i].Val = math.NaN()
+				}
+				outSeries = append(outSeries, serie2)
+				cache[Req{}] = append(cache[Req{}], serie2)
+				continue
+			}
+
+			for _, serie1 := range metaSeries[key] {
+				// no total
+				if _, ok := totalSeries[key]; !ok {
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,MISSING)", serie1.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,MISSING)", serie1.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = math.NaN()
+					}
+				} else {
+					serie2 := totalSeries[key]
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+					}
+				}
+				outSeries = append(outSeries, serie1)
+				cache[Req{}] = append(cache[Req{}], serie1)
+			}
+		}
+		return outSeries, nil
+	}
+
+	var totalsSerie models.Series
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalsSerie = sumSeries(series)
+		if len(series) == 1 {
+			totalsSerie.Target = fmt.Sprintf("sumSeries(%s)", totalsSerie.Target)
+			totalsSerie.QueryPatt = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.Tags = map[string]string{"name": totalsSerie.Target}
+		}
+	} else if totals != nil {
+		if len(totals) == 1 {
+			totalsSerie = totals[0]
+		} else if len(totals) == len(series) {
+			sort.Slice(series, func(i, j int) bool {
+				return series[i].Target < series[j].Target
+			})
+			sort.Slice(totals, func(i, j int) bool {
+				return totals[i].Target < totals[j].Target
+			})
+			for i, serie1 := range series {
+				serie2 := totals[i]
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
```

Comment:
I think this violates https://github.com/grafana/metrictank/blob/master/expr/NOTES#L35